### PR TITLE
Python: Dataflow, variable capture

### PR DESCRIPTION
**Pull request by yoff**
Tests showing that we do not handle variable capture (ported from C#, adapted slightly).

### CODE REVIEW

File: python/ql/test/experimental/dataflow/variable-capture/nonlocal.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        print("Unexpected flow", x)
+    else:
+        print("OK")
+
+
+def Out():
+    sinkO1 = ""
+    def captureOut1():
+        nonlocal sinkO1
+        sinkO1 = "source"
```
Comment (RasmusWL) [inline python/ql/test/experimental/dataflow/variable-capture/nonlocal.py:36]:
```suggestion
        sinkO1 = SOURCE
```

File: python/ql/test/experimental/dataflow/variable-capture/nonlocal.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    captureOut1()
+    SINK(sinkO1) #$ MISSING:captured
+
+    sinkO2 = ""
+    def captureOut2():
+        def m():
+            nonlocal sinkO2
+            sinkO2 = "source"
```
Comment (RasmusWL) [inline python/ql/test/experimental/dataflow/variable-capture/nonlocal.py:44]:
```suggestion
            sinkO2 = SOURCE
```

File: python/ql/test/experimental/dataflow/variable-capture/nonlocal.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        m()
+    captureOut2()
+    SINK(sinkO2) #$ MISSING:captured
+
+    nonSink0 = ""
+    def captureOut1NotCalled():
+        nonlocal nonSink0
+        nonSink0 = "source"
```
Comment (RasmusWL) [inline python/ql/test/experimental/dataflow/variable-capture/nonlocal.py:52]:
```suggestion
        nonSink0 = SOURCE
```

File: python/ql/test/experimental/dataflow/variable-capture/nonlocal.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    SINK_F(nonSink0)
+
+    def captureOut2NotCalled():
+        def m():
+            nonlocal nonSink0
+            nonSink0 = "source"
```
Comment (RasmusWL) [inline python/ql/test/experimental/dataflow/variable-capture/nonlocal.py:58]:
```suggestion
            nonSink0 = SOURCE
```

**Review by RasmusWL — CHANGES_REQUESTED** [review]
Overall LGTM, nice to have test of this :+1:

I would really like a comment (in the code) explaining why it is important to have BOTH `Out` and `Through` functions. From your explanation that variable-capture might end up depending on the call-graph, it makes sense to test both scenarios, but when I forget about this in 6 months time, I will certainly be confused about this subtle nuance.

Nitpick: what is up with calling the functions `Out` and not `out`? :smile:

File: python/ql/test/experimental/dataflow/coverage/argumentPassing.py [hunk 5]
```diff
@@ -112,7 +112,7 @@ def with_multiple_kw_args(a, b, c):
     SINK3(c)
 
 
-@expects(9)
+@expects(12)
```
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/test/experimental/dataflow/coverage/argumentPassing.py:115]: > was your test-setup broken? (I don't assume this is related to variable capture)
Indeed, an extra call was added some time ago.
 
> Should running these tests maybe be added as a step in our automated tests?
That would be very nice :-)

**Review by RasmusWL — APPROVED** [review]
Discussed in person that although I was not 100% convinced by the added comments explaining why we need both `out` and `through`, we would just move on :train: :shipit:

File: python/ql/test/experimental/dataflow/coverage/argumentPassing.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/test/experimental/dataflow/coverage/argumentPassing.py:115]: talked about this in person

**Pull request merged**